### PR TITLE
[JUJU-789] Proxy-aware charmhub HTTP client

### PR DESCRIPTION
**Pull request by SimonRichardson**
The following just adds a proxy-aware middleware to the charmhub HTTP
client. This seems like an oversight that the proxy information isn't
picked up. The new HTTP client has support for it out of the box, the
solution is to just add it.

## QA steps

```sh
$ juju bootstrap lxd test --model-default="https-proxy=<URL>"
$ juju deploy ubuntu
```

## Bug reference

https://bugs.launchpad.net/juju/+bug/1940327

### CODE REVIEW

**Review by juanmanuel-tirado — APPROVED** [review]
LG2M

File: apiserver/facades/client/charms/client.go [hunk 1]
```diff
@@ -603,19 +603,19 @@ func validateOrigin(origin params.CharmOrigin, schema string, switchCharm bool)
 }
 
 func (a *API) getCharmRepository(src corecharm.Source) (corecharm.Repository, error) {
-	a.mu.Lock()
```
Comment (hmlanigan) [inline apiserver/facades/client/charms/client.go:606]: Why is the lock being removed?

**Review by hmlanigan — APPROVED** [review]
I don't have a proxy setup to try, but the change didn't break non proxy behavior either.

File: apiserver/facades/client/charms/client.go (hunk 1 above)
Comment (SimonRichardson) [reply apiserver/facades/client/charms/client.go:606]: As we only trigger the nil check for the cached repoFactory with tests, it's not actually required anymore.

**Comment by SimonRichardson** [comment]
$$merge$$

**Pull request merged**